### PR TITLE
[flutter_tools] Add support for launching fuchsia app using session_control

### PR DESCRIPTION
**Pull request by michaellee8**
Allows fluttter app to be launched with session_control rather than tiles_ctl if such fuchsia build uses session_framework.

Further work on #83609. Refactored from #85450

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by michaellee8** [comment]
@zanderso @chaselatta I have factored out the changes related to session_control in #85450 into this PR. Please review. This should be merged before #85450.

File: packages/flutter_tools/lib/src/fuchsia/session_control.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    return result.exitCode == 0;
+  }
+
+  /// Instructs session_control on the device to launch the session at [url] as
+  /// an session.
+  ///
+  /// [url] should be formatted as a Fuchsia-style package URL, e.g.:
+  ///     fuchsia-pkg://fuchsia.com/flutter_gallery#meta/flutter_gallery.cmx
+  /// Returns true on success and false on failure.
+  Future<bool> launch(FuchsiaDevice device, String url) async {
```
Comment (chaselatta) [inline packages/flutter_tools/lib/src/fuchsia/session_control.dart:63]: We don't want to add this functionality to the flutter tool. Launching a session is only something that is needed for developers working on the session which is at a much lower level than a flutter application.

File: packages/flutter_tools/lib/src/fuchsia/session_control.dart [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    final RunResult result = await device.shell('session_control launch $url');
+    return result.exitCode == 0;
+  }
+
+  /// Instructs session_control on the device to restart.
+  ///
+  /// Returns true on success and false on failure.
+  Future<bool> restart(FuchsiaDevice device) async {
```
Comment (chaselatta) [inline packages/flutter_tools/lib/src/fuchsia/session_control.dart:71]: Same thing as above. We do not want the flutter tool to be restarting sessions.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart [hunk 3]
```diff
@@ -304,6 +324,12 @@ class FuchsiaDevice extends Device {
     bool ipv6 = false,
     String userIdentifier,
   }) async {
+    if (await isSession) {
+      globals.printTrace('Running on a workstation build.');
```
Comment (chaselatta) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:328]: please remove workstation here and replace with something about session_control.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart [hunk 4]
```diff
@@ -470,6 +514,15 @@ class FuchsiaDevice extends Device {
     covariant FuchsiaApp app, {
     String userIdentifier,
   }) async {
+    if (await isSession) {
+      // Currently there are no way to close a running app in ermine afaik.
+      // so we just restart the launcher to ensure that it is no longer running.
+      if (!await fuchsiaDeviceTools.sessionControl.restart(this)) {
```
Comment (chaselatta) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:520]: This is the equivalent to restarting you laptop every time you build. 

Can you remove this logic entirely? A user can close an application by clicking on the close button in ermine.

**Review by chaselatta — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart (hunk 4 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:520]: @chaselatta Out of curiosity, is there a `remove` to go along with the `add`? The Flutter CLI expects to be able to stop a Flutter app e.g. when a user types `q` in the terminal.

Comment (michaellee8) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:520]: @zanderso That was the problem. I had ro choose between no closing app snd force close by restart.

Comment (chaselatta) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:520]: Not right now but we can request that feature.

Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:520]: Got it. Yeah, in the meantime I agree with @chaselatta that this should be removed, and until there's a `remove` potentially added to `session_control` the user can just click the close button.

Comment (michaellee8) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:520]: > Got it. Yeah, in the meantime I agree with @chaselatta that this should be removed, and until there's a `remove` potentially added to `session_control` the user can just click the close button.

I too think so. Actually my initial implementation had a no-op stopApp but I just wanted to find a way to make restarting apps more roubust.

File: packages/flutter_tools/lib/src/fuchsia/session_control.dart (hunk 1 above)
Comment (michaellee8) [reply packages/flutter_tools/lib/src/fuchsia/session_control.dart]: Fixed in latest version.

**Comment by michaellee8** [comment]
@chaselatta Please review if there are any more problems.

**Comment by michaellee8** [comment]
@chaselatta Can you have a look at this? Thx.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart [hunk 5]
```diff
@@ -456,7 +500,7 @@ class FuchsiaDevice extends Device {
 
     // In a debug or profile build, try to find the observatory uri.
     final FuchsiaIsolateDiscoveryProtocol discovery =
-      getIsolateDiscoveryProtocol(appName);
+      getIsolateDiscoveryProtocol(fuchsiaUrl);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:503]: I don't understand this part of the change. What is the reason for changing this from the app name to the full url? Is there a unit test that can cover this change?

Comment (michaellee8) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart:503]: This is actually part of the change in observatory discovery. This should not be included in this patch. Previously the Observatory discovery prototool identify the app we are running by connecting to vmservice of each running app component and then compare each name of flutter isolaten with appName, the new implementation does all discovery based on /hub and comparing the component url. More explanation are written in #85450.

Right now I should stick with appName on this PR. Let me change it soon.

**Review by chaselatta — APPROVED** [review]
LGTM once Zach's comments are addressed.

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart (hunk 5 above)
Comment (michaellee8) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart]: @zanderso Please review again. I have revertd those changes.

**Comment by michaellee8** [comment]
I cannot understand why the CI fails on customer test. Did they changed ci config?

**Review by zanderso — APPROVED** [review]

**Pull request merged**